File: NET_4.5.2/C1.WPF.FlexGrid/VB/UnboundCellFactoryWPF/UnboundCellFactory.vb
Imports System.Windows
Imports System.Windows.Media
Imports System.Windows.Controls
Imports C1.WPF.FlexGrid
Imports C1.WPF.Extended

Public Class UnboundCellFactory
	Inherits CellFactory
	Private _flex As C1FlexGrid

	' ctor
	Public Sub New(flex As C1FlexGrid)
		_flex = flex
		AddHandler _flex.ScrollPositionChanging, AddressOf flex_ScrollPositionChanging
		AddHandler _flex.BeginningEdit, AddressOf flex_BeginningEdit
	End Sub

	' finish edits before scrolling
	Private Sub flex_ScrollPositionChanging(sender As Object, e As EventArgs)
		_flex.FinishEditing()
	End Sub

	' cancel standard editing for our custom cells (they are editors already)
	Private Sub flex_BeginningEdit(sender As Object, e As CellEditEventArgs)
        Dim t As Type = GetCellType(e.CellRange)
		If t IsNot Nothing Then
			e.Cancel = True
		End If
	End Sub

	' create the custom cell editor
	Public Overrides Sub CreateCellContent(grid As C1FlexGrid, bdr As Border, rng As CellRange)
        Dim t As Type = GetCellType(rng)
		If t IsNot Nothing Then
			' cell has a custom type, create editor
			bdr.Child = CreateEditor(t, rng)
		Else
            ' default handling
			MyBase.CreateCellContent(grid, bdr, rng)
		End If
	End Sub

	' select the element type for a cell
	Private Function GetCellType(rng As CellRange) As Type
        Dim c As Column = _flex.Columns(rng.Column)
		If c.ColumnName = "Data" Then
			Dim r = _flex.Rows(rng.Row)
			Return TryCast(r.Tag, Type)
		End If
		Return Nothing
	End Function

	' initialize custom cell editor
	Private Function CreateEditor(t As Type, rng As CellRange) As UIElement
		' create cell element, save range in Tag property
        Dim edt As FrameworkElement = TryCast(Activator.CreateInstance(t), FrameworkElement)
		edt.Tag = rng
		edt.VerticalAlignment = VerticalAlignment.Center

		' get current value from grid
        Dim value As Object = _flex(rng.Row, rng.Column)

		' initialize each editor type
        Dim textBox As TextBox = TryCast(edt, TextBox)
		If textBox IsNot Nothing Then
			textBox.Text = If(TypeOf value Is String, DirectCast(value, String), String.Empty)
			AddHandler textBox.LostFocus, AddressOf StoreCellValue
		End If

        Dim checkBox As CheckBox = TryCast(edt, CheckBox)
		If checkBox IsNot Nothing Then
			checkBox.HorizontalAlignment = HorizontalAlignment.Center
			checkBox.VerticalAlignment = VerticalAlignment.Center
			checkBox.IsChecked = If(TypeOf value Is Boolean, CBool(value), False)
			AddHandler checkBox.LostFocus, AddressOf StoreCellValue
		End If

        Dim comboBox As ComboBox = TryCast(edt, ComboBox)
		If comboBox IsNot Nothing Then
			For i As Integer = 0 To 4
				comboBox.Items.Add("Item " & i.ToString())
			Next
			comboBox.SelectedValue = If(TypeOf value Is String, DirectCast(value, String), comboBox.Items(0))
			AddHandler comboBox.LostFocus, AddressOf StoreCellValue
		End If

        Dim slider As Slider = TryCast(edt, Slider)
		If slider IsNot Nothing Then
			slider.Minimum = 0
			slider.Maximum = 50
			slider.Value = If(TypeOf value Is Double, CDbl(value), 0)
			AddHandler slider.LostFocus, AddressOf StoreCellValue
		End If

        Dim colorPicker As C1ColorPicker = TryCast(edt, C1ColorPicker)
		If colorPicker IsNot Nothing Then
			colorPicker.SelectedColor = If(TypeOf value Is Color, CType(value, Color), Colors.White)
			AddHandler colorPicker.SelectedColorChanged, AddressOf StoreCellValue
		End If

		' done
		edt.VerticalAlignment = VerticalAlignment.Center
		Return edt
	End Function

	' finalize custom cell editor
	Public Overrides Sub DisposeCell(grid As C1FlexGrid, cellType As CellType, cell As FrameworkElement)
        Dim edt As FrameworkElement = TryCast(DirectCast(cell, Border).Child, FrameworkElement)
		If edt IsNot Nothing Then
            Dim colorPicker As C1ColorPicker = TryCast(edt, C1ColorPicker)
			If colorPicker IsNot Nothing Then
				RemoveHandler colorPicker.SelectedColorChanged, AddressOf StoreCellValue
			Else
				RemoveHandler edt.LostFocus, AddressOf StoreCellValue
			End If
		End If
		MyBase.DisposeCell(grid, cellType, cell)
	End Sub

    ' store edited value into grid cell when the editor loses focus/changes
	Private Sub StoreCellValue(sender As Object, e As EventArgs)
		' get editor and range
        Dim edt As FrameworkElement = TryCast(sender, FrameworkElement)
        Dim rng As CellRange = CType(edt.Tag, CellRange)

		' apply value
        Dim textBox As TextBox = TryCast(edt, TextBox)
		If textBox IsNot Nothing Then
			_flex(rng.Row, rng.Column) = textBox.Text
		End If
        Dim checkBox As CheckBox = TryCast(edt, CheckBox)
		If checkBox IsNot Nothing Then
			_flex(rng.Row, rng.Column) = checkBox.IsChecked
		End If

        Dim comboBox As ComboBox = TryCast(edt, ComboBox)
		If comboBox IsNot Nothing Then
			_flex(rng.Row, rng.Column) = comboBox.SelectedValue
		End If

        Dim slider As Slider = TryCast(edt, Slider)
		If slider IsNot Nothing Then
			_flex(rng.Row, rng.Column) = slider.Value
		End If

        Dim colorPicker As C1ColorPicker = TryCast(edt, C1ColorPicker)
		If colorPicker IsNot Nothing Then
			_flex(rng.Row, rng.Column) = colorPicker.SelectedColor
		End If
	End Sub
End Class

File: NET_4.5.2/C1.WPF.FlexGrid/CS/UnboundCellFactoryWPF/UnboundCellFactory.cs
using System;
using System.Windows;
using System.Windows.Media;
using System.Windows.Controls;
using C1.WPF.FlexGrid;
using C1.WPF.Extended;

namespace UnboundCellFactoryWPF
{
    public class UnboundCellFactory : CellFactory
    {
        C1FlexGrid _flex;

        // ctor
        public UnboundCellFactory(C1FlexGrid flex)
        {
            _flex = flex;
            _flex.ScrollPositionChanging += flex_ScrollPositionChanging;
            _flex.BeginningEdit += flex_BeginningEdit;
        }

        // finish edits before scrolling
        void flex_ScrollPositionChanging(object sender, EventArgs e)
        {
            _flex.FinishEditing();
        }

        // cancel standard editing for our custom cells (they are editors already)
        void flex_BeginningEdit(object sender, CellEditEventArgs e)
        {
            var t = GetCellType(e.CellRange);
            if (t != null)
            {
                e.Cancel = true;
            }
        }

        // create the custom cell editor
        public override void CreateCellContent(C1FlexGrid grid, Border bdr, CellRange rng)
        {
            var t = GetCellType(rng);
            if (t != null)
            {
                // cell has a custom type, create editor
                bdr.Child = CreateEditor(t, rng);
            }
            else
            {
                // default handling
                base.CreateCellContent(grid, bdr, rng);
            }
        }

        // select the element type for a cell
        Type GetCellType(CellRange rng)
        {
            var c = _flex.Columns[rng.Column];
            if (c.ColumnName == "Data")
            {
                var r = _flex.Rows[rng.Row];
                return r.Tag as Type;
            }
            return null;
        }

        // initialize custom cell editor
        UIElement CreateEditor(Type t, CellRange rng)
        {
            // create cell element, save range in Tag property
            var edt = Activator.CreateInstance(t) as FrameworkElement;
            edt.Tag = rng;
            edt.VerticalAlignment = VerticalAlignment.Center;

            // get current value from grid
            var value = _flex[rng.Row, rng.Column];

            // initialize each editor type
            var textBox = edt as TextBox;
            if (textBox != null)
            {
                textBox.Text = value is string ? (string)value : string.Empty;
                textBox.LostFocus += StoreCellValue;
            }

            var checkBox = edt as CheckBox;
            if (checkBox != null)
            {
                checkBox.HorizontalAlignment = HorizontalAlignment.Center;
                checkBox.VerticalAlignment = VerticalAlignment.Center;
                checkBox.IsChecked = value is bool ? (bool)value : false;
                checkBox.LostFocus += StoreCellValue;
            }

            var comboBox = edt as ComboBox;
            if (comboBox != null)
            {
                for (int i = 0; i < 5; i++)
                {
                    comboBox.Items.Add("Item " + i.ToString());
                }
                comboBox.SelectedValue = value is string ? (string)value : comboBox.Items[0];
                comboBox.LostFocus += StoreCellValue;
            }

            var slider = edt as Slider;
            if (slider != null)
            {
                slider.Minimum = 0;
                slider.Maximum = 50;
                slider.Value = value is double ? (double)value : 0;
                slider.LostFocus += StoreCellValue;
            }

            var colorPicker = edt as C1ColorPicker;
            if (colorPicker != null)
            {
                colorPicker.SelectedColor = value is Color ? (Color)value : Colors.White;
                colorPicker.SelectedColorChanged += StoreCellValue;
            }

            // done
            edt.VerticalAlignment = VerticalAlignment.Center;
            return edt;
        }

        // finalize custom cell editor
        public override void DisposeCell(C1FlexGrid grid, CellType cellType, FrameworkElement cell)
        {
            var edt = ((Border)cell).Child as FrameworkElement;
            if (edt != null)
            {
                var colorPicker = edt as C1ColorPicker;
                if (colorPicker != null)
                {
                    colorPicker.SelectedColorChanged -= StoreCellValue;
                }
                else
                {
                    edt.LostFocus -= StoreCellValue;
                }
            }
            base.DisposeCell(grid, cellType, cell);
        }

        // store edited value into grid cell when the editor loses focus/changes
        void StoreCellValue(object sender, EventArgs e)
        {
            // get editor and range
            var edt = sender as FrameworkElement;
            var rng = (CellRange)edt.Tag;

            // apply value
            var textBox = edt as TextBox;
            if (textBox != null)
            {
                _flex[rng.Row, rng.Column] = textBox.Text;
            }
            var checkBox = edt as CheckBox;
            if (checkBox != null)
            {
                _flex[rng.Row, rng.Column] = checkBox.IsChecked;
            }

            var comboBox = edt as ComboBox;
            if (comboBox != null)
            {
                _flex[rng.Row, rng.Column] = comboBox.SelectedValue;
            }

            var slider = edt as Slider;
            if (slider != null)
            {
                _flex[rng.Row, rng.Column] = slider.Value;
            }

            var colorPicker = edt as C1ColorPicker;
            if (colorPicker != null)
            {
                _flex[rng.Row, rng.Column] = colorPicker.SelectedColor;
            }
        }
    }
}